### PR TITLE
Add MangaTerra

### LINKED ISSUE #1761: Manga Terra: Source Request
### Source name

Manga Terra

### Source link

https://manga-terra.com

### Source language

Brazilian Portuguese

### Other details

NSFW = true

### Acknowledgements

- [X] I have checked that the extension does not already exist by searching the [extension listing](https://keiyoushi.github.io/extensions/) and verified it does not appear there.
- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a meaningful title with the source name.
- [ ] I have admitted that I am a clown by having checked this box, as I have not read these acknowledgements.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by choppeh**
Closes #1761

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

### CODE REVIEW

File: src/pt/mangaterra/AndroidManifest.xml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+            android:theme="@android:style/Theme.NoDisplay">
+            <intent-filter>
+                <action android:name="android.intent.action.VIEW" />
+
+                <category android:name="android.intent.category.DEFAULT" />
+                <category android:name="android.intent.category.BROWSABLE" />
+
+                <data
+                    android:scheme="https"
+                    android:host="manga-terra.com/"
```
Comment (vetleledaal) [inline src/pt/mangaterra/AndroidManifest.xml]:
```suggestion
                    android:host="manga-terra.com"
```

File: src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        description = document.selectFirst(".card-body p")?.ownText()
+        thumbnail_url = document.selectFirst(".card-body img")?.srcAttr()
+        genre = document.select(".card-series-about a").joinToString { it.ownText() }
+    }
+
+    override fun pageListParse(document: Document): List<Page> {
+        val pages = mutableListOf<Page>()
+        var page = 1
+        do {
+            val dom = client.newCall(GET("${document.location()}/${page++}", headers)).execute().asJsoup()
```
Comment (vetleledaal) [inline src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt]: You shouldn't make a new call here for each page, use `imageUrlParse`. If the total page count is unknown you could do a binary search -- could still be slow with the current rate limit.

You can check [Akuma](https://github.com/keiyoushi/extensions-source/blob/2fd542b42b3ab81587bc37568976bc9c8034906c/src/all/akuma/src/eu/kanade/tachiyomi/extension/all/akuma/Akuma.kt#L207) for imageUrlParse
You can check [some older version of MangaHub](https://github.com/kevin01523/tachiyomi-extensions/blob/d292dac0418b4bc90864a54b60c3d7a1c736ead4/multisrc/src/main/java/eu/kanade/tachiyomi/multisrc/mangahub/MangaHub.kt#L255) for binary search

File: src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        if (response.request.url.pathSegments.contains("search")) {
+            return searchByQueryMangaParse(response)
+        }
+        return super.searchMangaParse(response)
+    }
+
+    override fun searchMangaNextPageSelector() = popularMangaNextPageSelector()
+
+    override fun fetchSearchManga(page: Int, query: String, filters: FilterList): Observable<MangasPage> {
+        if (query.startsWith(slugPrefix)) {
```
Comment (vetleledaal) [inline src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt:107]: Throws `kotlin.UninitializedPropertyAccessException` for `url` when manually searching for `slug:martial-peak`.

Non-existing slug throws `NullPointerException`, but that's probably fine.

File: src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    override fun popularMangaSelector(): String = ".card-body .row > div"
+
+    override fun searchMangaFromElement(element: Element) = popularMangaFromElement(element)
+
+    private fun searchByQueryMangaParse(response: Response): MangasPage {
+        val fragment = Jsoup.parse(
+            json.decodeFromString<String>(response.body.string()),
+            baseUrl,
+            Parser.htmlParser(),
+        )
```
Comment (AwkwardPeak7) [inline src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt]:
```suggestion
        val fragment = Jsoup.parseBodyFragment(
            json.decodeFromString<String>(response.body.string()),
            baseUrl,
        )
```

File: src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt
Comment (AwkwardPeak7) [inline src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt:1]: Please consider grouping methods based on their functionality. Like all popular methods together, then all latest methods and so on

**Review by AwkwardPeak7 — CHANGES_REQUESTED** [review]

File: src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    override fun latestUpdatesRequest(page: Int) = GET("$baseUrl/manga?q=u&page=$page", headers)
+
+    override fun latestUpdatesSelector() = popularMangaSelector()
+
+    override fun mangaDetailsParse(document: Document) = SManga.create().apply {
+        title = document.selectFirst(".card-body h1")!!.ownText()
+        description = document.selectFirst(".card-body p")?.ownText()
+        thumbnail_url = document.selectFirst(".card-body img")?.srcAttr()
+        genre = document.select(".card-series-about a").joinToString { it.ownText() }
+        setUrlWithoutDomain(document.location())
```
Comment (AwkwardPeak7) [inline src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt:74]:
```suggestion
```

File: src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    override fun getFilterList(): FilterList {
+        CoroutineScope(Dispatchers.IO).launch { fetchGenres() }
+        val filters = mutableListOf(
+            GenreFilter(
+                title = "Categorias",
+                genres = genresList,
+            ),
+        )
+        return FilterList(filters)
+    }
```
Comment (AwkwardPeak7) [inline src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt]: add a warning/note to tell user to press reset to load the genrelist. also don't add when it is empty

File: src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+    private fun findPageCount(pageUrl: String): Int {
+        var lowerBound = 1
+        var upperBound = 100
+
+        val client = network.client.newBuilder()
+            .followRedirects(false)
+            .rateLimit(1, 2, TimeUnit.SECONDS)
+            .build()
```
Comment (AwkwardPeak7) [inline src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt]: save this outside the function as `noRedirectClient`

**Review by AwkwardPeak7 — CHANGES_REQUESTED** [review]

File: src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt (hunk 5 above)
Comment (choppeh) [reply src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt:74]: Method call required when searching with ```slug:<manga-id>``` in ```fetchSearchManga```. 

See: https://github.com/keiyoushi/extensions-source/pull/2300/files#r1559936766

But I can change it to set the ``url`` with ``apply`` inline. If that's really the best way.

Comment (AwkwardPeak7) [reply src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt:74]: Ah, it's fine then

File: src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+    override val lang: String = "pt-BR"
+    override val supportsLatest: Boolean = true
+    override val name: String = "Manga Terra"
+    override val baseUrl: String = "https://manga-terra.com"
+
+    override val client = network.client.newBuilder()
+        .rateLimit(1, 2, TimeUnit.SECONDS)
+        .build()
+
+    private val noRedirectClient = network.client.newBuilder()
```
Comment (AwkwardPeak7) [inline src/pt/mangaterra/src/eu/kanade.tachiyomi.extension.pt.mangaterra/MangaTerra.kt]: use `cloudflareClient`

**Review by AwkwardPeak7 — APPROVED** [review]

**Pull request merged**